### PR TITLE
enhance: [2.4] Reduce segmentManager lock granularity

### PR DESCRIPTION
**Pull request by bigsheeper**
Use a channel level key lock for segments in segmentManager.

issue: https://github.com/milvus-io/milvus/issues/37633, https://github.com/milvus-io/milvus/issues/37630

pr: https://github.com/milvus-io/milvus/pull/37836

### CODE REVIEW

File: internal/datacoord/segment_manager.go [hunk 1]
```diff
@@ -435,30 +439,44 @@ func (s *SegmentManager) DropSegment(ctx context.Context, segmentID UniqueID) {
...
+
+	sealed, _ := s.channel2Sealed.GetOrInsert(channel, typeutil.NewUniqueSet())
+	growing, ok := s.channel2Growing.Get(channel)
+	if !ok {
+		return sealed.Collect(), nil
 	}
 
-	sealedSegments := s.meta.GetSegments(segCandidates, func(segment *SegmentInfo) bool {
-		return segment.CollectionID == collectionID && isSegmentHealthy(segment) && segment.State == commonpb.SegmentState_Sealed
+	sealedSegments := s.meta.GetSegments(sealed.Collect(), func(segment *SegmentInfo) bool {
```
Comment (czs007) [inline internal/datacoord/segment_manager.go:455]: Move this to the 'else' branch of 'if len(segIDs) != 0'

**Comment by bigsheeper** [comment]
rerun ut

**Pull request merged**